pkg/ingester/stream_test.go

```diff
@@ -149,6 +151,109 @@ func TestPushDeduplication(t *testing.T) {
 	require.Equal(t, len("test"+"newer, better test"), written)
 }
 
+func TestPushDoesNotDedupeWhenMetadataDiffers(t *testing.T) {
+	limits, err := validation.NewOverrides(defaultLimitsTestConfig(), nil)
+	require.NoError(t, err)
+	limiter := NewLimiter(limits, NilMetrics, newIngesterRingLimiterStrategy(&ringCountMock{count: 1}, 1), &TenantBasedStrategy{limits: limits})
+	retentionHours := util.RetentionHours(limiter.limits.RetentionPeriod("fake"))
+	// Force structured-metadata-capable format
+	chunkfmt, headfmt := chunkenc.ChunkFormatV4, chunkenc.UnorderedWithStructuredMetadataHeadBlockFmt
+
+	s := newStream(
+		chunkfmt,
+		headfmt,
+		defaultConfig(),
+		limiter.rateLimitStrategy,
+		"fake",
+		model.Fingerprint(0),
+		labels.FromStrings("foo", "bar"),
+		true,
+		NewStreamRateCalculator(),
+		NilMetrics,
+		nil,
+		nil,
+		retentionHours,
+		noPolicy,
+	)
+
+	// Two entries with same ts and line, but different StructuredMetadata
+	entries := []logproto.Entry{
+		{Timestamp: time.Unix(1, 0), Line: "same", StructuredMetadata: push.LabelsAdapter{push.LabelAdapter{Name: "k", Value: "a"}}},
+		{Timestamp: time.Unix(1, 0), Line: "same", StructuredMetadata: push.LabelsAdapter{push.LabelAdapter{Name: "k", Value: "b"}}},
+	}
+
+	_, err = s.Push(context.Background(), entries, recordPool.GetRecord(), 0, true, false, nil, "loki")
+	require.NoError(t, err)
+
+	// Desired behavior: both entries should be stored (this will fail today)
+	require.Len(t, s.chunks, 1)
+	require.Equal(t, 2, s.chunks[0].chunk.Size(),
+		"expected no dedupe when StructuredMetadata differs, but dedupe occurred")
+}
+
+func TestEndToEnd_Dedup_WithStructuredMetadata(t *testing.T) {
+	limits, err := validation.NewOverrides(defaultLimitsTestConfig(), nil)
+	require.NoError(t, err)
+	limiter := NewLimiter(limits, NilMetrics, newIngesterRingLimiterStrategy(&ringCountMock{count: 1}, 1), &TenantBasedStrategy{limits: limits})
+	retentionHours := util.RetentionHours(limiter.limits.RetentionPeriod("fake"))
+	// Force structured-metadata-capable format
+	chunkfmt, headfmt := chunkenc.ChunkFormatV4, chunkenc.UnorderedWithStructuredMetadataHeadBlockFmt
+
+	// Ensure we run this for unordered writes to exercise the unordered head block path which performs its own dedupe.
+	s := newStream(
+		chunkfmt,
+		headfmt,
+		defaultConfig(),
+		limiter.rateLimitStrategy,
+		"fake",
+		model.Fingerprint(0),
+		labels.FromStrings("foo", "bar"),
+		true,
+		NewStreamRateCalculator(),
+		NilMetrics,
+		nil,
+		nil,
+		retentionHours,
+		noPolicy,
+	)
+
+	// Push two entries with identical ts and line but different StructuredMetadata
+	baseTS := time.Unix(100, 0)
+	entries := []logproto.Entry{
```

Comment:
could you add a 3rd entry that does match to make sure that logic still works as expected?